Implement Delta Scores in search tasks

zrolfs opened the pull request:
Preliminary Results (before some fixes, update incoming):
Using Delta Scores resulted in an average PSM increase of 3.6%, and cannot do any harm to the accuracy because of the way in which they have been implemented (see fraction 8 & 27). There is a mild (9%) increase in search time. Data shown below is from the 28 jurkat trypsin fractions.
![image](https://user-images.githubusercontent.com/16883585/38166859-68fe870c-34f0-11e8-9b82-4cfdaf3eaa47.png)

zrolfs commented:
Searches are a little slower regardless of if delta scores are being calculated, as the runner up is still being saved in most cases. After fixes, total PSMs from the fractions went up from 276368 => 289859, a 4.9% increase.

zrolfs commented:
Just had a thought: this should probably hold off until we've localized immediately after GPTMD, otherwise the ambiguity between reside positions will hurt identification of discovered modifications.

trishorts commented:
can we do percolator on these two sets and see how they fair afterward? maybe just try fractions 14 and 15?

trishorts commented:
Is it safe to say that delta score is analogous to E-value except that you only look at the closes score and not the distribution of next best scores? Along those lines i'm wondering how many things didn't have any other option in the database with the same mass to pick from. In that case, you get a set of guaranteed winners. That could be good and bad. I think that I've argued that the bar should be lowered for peptides without much competition. On the other hand, those winning a highly competitive battle might be stronger hits. This is all very interesting.

rmillikin commented:
I have to say, I don't want this to be default behavior, at least right now. Messing with the default scoring scheme is a huge change and with HUPO, Stefan's paper, Lei's paper, and ASMS coming up I think it is not a good idea to change this now. This kind of change should undergo a lot of internal testing before we release it for general use. I'm OK with it being under an "advanced options" section that is unchecked by default.

zrolfs commented:
@trishorts I think E-value and delta-score are similar in that they are both comparing the top score against other scores, but the meaning behind them is different. E-value is aimed at determining the probability that that score was randomly obtained (good at knocking down noisy/messy spectra), while delta-score is aimed at catching mis-identifications caused by ambiguity. From our work with E-values, I recall very few spectra that had no other options in the database. The current default for delta-values is that anything without a "runner up" score is assigned the score cutoff (Default=5) as its runner up. I think those winning a highly competitive battle (ie 24 vs 27) have a high chance of being incorrect. This is problematic for peptides with multiple modification sites, however. Or do you mean PSMs searched against many theoreticals (regardless of scores) as being competitive?

@rmillikin I've changed the default for delta scores to be off. There is still a slight increase in search time, however.
I also changed the default minimum peptide length from 5 to 7. @rmmiller22 had a comment at her HUPO talk about small peptide lengths being suspicious, and I agree. A 5-mer has a max score of 7, which is always in our noise. 6-mer has a max score of 9 assuming all fragments are present. 7-mer is slightly more realistic. An arguement was made at HUPO for 9-mers, but I think that's too far. Baby steps first.

zrolfs commented on EngineLayer/FdrAnalysis/FdrAnalysisResults.cs:
```diff
@@ -25,6 +27,7 @@ public override string ToString()
             var sb = new StringBuilder();
             sb.AppendLine(base.ToString());
             sb.Append("PSMs within 1% fdr: " + PsmsWithin1PercentFdr);
+            sb.Append("Delta Score Improvement: " + DeltaScoreImprovement);
```
This is a bool that says if delta score was used. It is always false when useDeltaScoresForFdr is false. It is not always true when useDeltaScoresForFdr is true, however.

zrolfs replied:
I would like to keep it as confirmation that delta scores was not used when useDeltaScoresForFdr is unchecked.

rmillikin replied:
ok, that sounds good. Please just give a little more info in the text what it means
